pyomo/core/tests/unit/test_expr_pyomo5.py

```diff
@@ -6379,6 +6379,355 @@ def test_replace_nested(self):
+            else:
+                return [node]
+        walker = EXPR.GenericExpressionVisitor(exitNode=exit)
+        m = self.m
+
+        ans = walker.walk_expression(self.e)
+        ref = [
+            [[m.x], [2]],
+            [m.y],
+            [[m.z], [[m.x], [m.y]]]
+        ]
+        self.assertEqual(str(ans), str(ref))
+
+        ans = walker.walk_expression(m.x)
+        ref = [m.x]
+        self.assertEqual(str(ans), str(ref))
+
+        ans = walker.walk_expression(2)
+        ref = [2]
+        self.assertEqual(str(ans), str(ref))
+
+    def test_call_on_nonexpression(self):
+        def enter(node):
+            if type(node) in nonpyomo_leaf_types \
+               or not node.is_expression_type():
+                return (), node
+            return node.args, []
+        def finalize(result):
+            if type(result) is list:
+                return result
+            else:
+                return [result]
+        walker = EXPR.GenericExpressionVisitor(
+            enterNode=enter, finalizeResult=finalize)
+
+    def test_beforeChild_acceptData_afterChild(self):
+        counts = [0,0,0]
+        def before(node, child):
+            counts[0] += 1
+            if type(child) in nonpyomo_leaf_types \
+               or not child.is_expression_type():
+                return False, child
+        def accept(node, data, child_result):
+            counts[1] += 1
+        def after(node, child):
+            counts[2] += 1
+        walker = EXPR.GenericExpressionVisitor(
+            beforeChild=before, acceptData=accept, afterChild=after)
+        ans = walker.walk_expression(self.e)
+        m = self.m
+        ref = []
+        self.assertEqual(ans, ref)
+        self.assertEquals(counts, [9,9,9])
+
+    def test_enterNode(self):
+        ans = []
+        def before(node, child):
+            if type(child) in nonpyomo_leaf_types \
+               or not child.is_expression_type():
+                return False, child
+        def accept(node, data, child_result):
+            if data is not child_result:
+                data.append(child_result)
+        def enter(node):
+            if type(node) in nonpyomo_leaf_types \
+               or not node.is_expression_type():
+                return (), ans
+            else:
+                return node.args, ans
+        walker = EXPR.GenericExpressionVisitor(
+            enterNode=enter, beforeChild=before, acceptData=accept)
+        ans = walker.walk_expression(self.e)
+        m = self.m
+        ref = [m.x, 2, m.y, m.z, m.x, m.y]
+        self.assertEqual(str(ans), str(ref))
+
+    def test_finalize(self):
+        ans = []
+        def before(node, child):
+            if type(child) in nonpyomo_leaf_types \
+               or not child.is_expression_type():
+                return False, child
+        def accept(node, data, child_result):
+            if data is not child_result:
+                data.append(child_result)
+        def enter(node):
```

Comment:
Same is true here as for above test... before should make this first if not ever true.